diff -urN wmconfig-0.9.10.orig/input-gnome.c wmconfig-0.9.10/input-gnome.c
--- wmconfig-0.9.10.orig/input-gnome.c	2003-07-31 12:34:56.000000000 +0200
+++ wmconfig-0.9.10/input-gnome.c	2003-07-31 12:30:33.000000000 +0200
@@ -33,7 +33,7 @@
 
 #include "wmconfig.h"
 
-#define TERMINAL_USED "xterm -e"
+#define TERMINAL_USED "xterm"
 
 /*
  * This reads & parses a system configuration file
@@ -90,9 +90,13 @@
 
 	if (! add_for_terminal && dentry->terminal) {
 		add_for_terminal = strlen(TERMINAL_USED);
-		exec = realloc(exec, strlen(exec)+add_for_terminal+2);
+		exec = realloc(exec, strlen(exec)+2*add_for_terminal+strlen(dentry->exec[i])+13);
 		strcat(exec, TERMINAL_USED);
-		strcat(exec, " ");
+		strcat(exec, " -name ");
+		strcat(exec, TERMINAL_USED);
+		strcat(exec, ".");
+		strcat(exec, dentry->exec[i]);
+		strcat(exec, " -e ");
 	}
 
 	exec = realloc(exec, strlen(exec)+strlen(dentry->exec[i])+2);
