diff -urN wmconfig-0.9.10.orig/Makefile.am wmconfig-0.9.10/Makefile.am
--- wmconfig-0.9.10.orig/Makefile.am	Sat Aug 28 00:38:16 1999
+++ wmconfig-0.9.10/Makefile.am	Fri Oct 12 16:43:28 2001
@@ -13,7 +13,7 @@
 man_MANS	= wmconfig.1
 
 if HAVE_GNOME
-EXTRA_FILES	+= input-gnome.c
+EXTRA_FILES	= input-gnome.c
 EXTRA_INCLUDES	= @GNOME_INCS@
 EXTRA_LIBS      = @GNOME_LIBS@
 else
@@ -22,7 +22,7 @@
 EXTRA_LIBS      =
 endif
 
-INCLUDES	+= $(EXTRA_INCLUDES)
+INCLUDES	= $(EXTRA_INCLUDES)
 WM_FILES	= wmconfig.c input.c rootmenu.c afterstep_dir.c kde.c wmaker.c \
 		  package.c util.c afterstep.c fvwm95.c mwm.c debug.c fvwm2.c \
 		  icewm.c blackbox.c parse.c
