diff -urN wmconfig-0.9.10/gnome-dentry.c wmconfig-0.9.10.new/gnome-dentry.c
--- wmconfig-0.9.10/gnome-dentry.c	2003-06-23 23:29:57.000000000 +0200
+++ wmconfig-0.9.10.new/gnome-dentry.c	2003-06-23 23:25:08.000000000 +0200
@@ -265,6 +265,10 @@
 				}
 			} else {
 				newitem->icon = gnome_pixmap_file (icon_base);
+				if (!newitem->icon)
+					/* Maybe it didn't have an extension? */
+					newitem->icon = gnome_pixmap_file (g_strconcat(icon_base, ".xpm", NULL));
+
 				/* If it wasn't found, look in kde */
 				if (!newitem->icon) {
 					gchar *iconname = g_concat_dir_and_file (KDE_ICONDIR "/hicolor/32x32/apps", icon_base);
diff -urN wmconfig-0.9.10/icewm.c wmconfig-0.9.10.new/icewm.c
--- wmconfig-0.9.10/icewm.c	2003-06-23 22:53:44.000000000 +0200
+++ wmconfig-0.9.10.new/icewm.c	2003-06-23 22:55:41.000000000 +0200
@@ -66,7 +66,7 @@
                 for (sp = 1; sp <= level; sp++)
 		    putchar('\t');
 		printf("prog \"%s\" %s ",
-                       unicode_to_locale(app->name), app->__package);
+                       unicode_to_locale(app->name), strcmp(app->icon,"")?app->icon:"\"\"");
 					
                 /* this removes ending '&' that should not be there */
                 {
diff -urN wmconfig-0.9.10/input-gnome.c wmconfig-0.9.10.new/input-gnome.c
--- wmconfig-0.9.10/input-gnome.c	2003-06-23 22:53:44.000000000 +0200
+++ wmconfig-0.9.10.new/input-gnome.c	2003-06-23 22:52:38.000000000 +0200
@@ -41,7 +41,7 @@
 int read_gnome_file(const char *filename, const char *group)
 {
     GnomeDesktopEntry *dentry;
-    char               *package, *name, *exec;
+    char               *package, *name, *exec, *icon;
     int                i, add_for_terminal;
 
 /*    fprintf(stderr, "scanning file %s in group %s\n",filename, group); */
@@ -77,6 +77,8 @@
     } else if (dentry->exec[0]) {
 	package = dentry->exec[0];
     }
+    if (dentry->icon)
+    	icon = dentry->icon;
 
     name = package;
     
@@ -101,6 +103,7 @@
     exec = realloc(exec, strlen(exec)+2);
     strcat(exec, "&");
     update_package(package, "exec", exec);
+    update_package(package, "icon", icon);
     update_package(package, "name", name);
     update_package(package, "group", group);
 
